torch/csrc/jit/fuser/executor.cpp

```diff
@@ -358,34 +370,191 @@ bool runFusion(const int64_t key, Stack& stack) {
+            if (target_size.size() != size_to_sum_to.size()) {
+              // we have dimensions to delete...
+              needs_sumtosize_other_dim = true;
+              dims_to_go = size_to_sum_to.size() - target_size.size();
+              AT_ASSERTM(dims_to_go > 0, "inconsistent concat sizes in fusion");
+              concat_dim = c.dim() - dims_to_go;
+              for (size_t d = 0; d < dims_to_go; d++) {
+                size_to_sum_to[d] = 1;
+              }
+            }
+            for (size_t d = 0; d < target_size.size(); d++) {
+              if (d != concat_dim) {
+                needs_sumtosize_other_dim |=
+                    (size_to_sum_to[d + dims_to_go] != target_size[d]);
+                size_to_sum_to[d + dims_to_go] = target_size[d];
+              } else {
+                needs_sumtosize_concat_dim |=
+                    (sizes_concat_dim[0] != target_size[d]);
+                sizes_concat_dim[0] = target_size[d];
+              }
+            }
+          }
+
+          // check the others for compatibility in other dimensions
+          // and scan the target sizes of the concat dimension
+          for (size_t i = 1; i < size_inputs.size(); i++) {
+            if (size_inputs[i] == -1) {
+              AT_ASSERTM(
+                  !needs_sumtosize_other_dim,
+                  "inconsistent concat sizes in fusion");
+            } else {
+              auto target_size = all_inputs[size_inputs[i]].toIntListRef();
+              AT_ASSERTM(
+                  target_size.size() + dims_to_go == size_to_sum_to.size(),
+                  "inconsistent concat sizes in fusion");
+              for (size_t d = 0; d < target_size.size(); d++) {
+                if (d == concat_dim) {
+                  needs_sumtosize_concat_dim |=
+                      (sizes_concat_dim[i] != target_size[d]);
+                  sizes_concat_dim[i] = target_size[d];
+                } else {
+                  // other dimensions must match first tensors example
+                  AT_ASSERTM(
+                      size_to_sum_to[d + dims_to_go] == target_size[d],
+                      "inconsistent concat sizes in fusion");
+                }
+              }
+            }
+          }
+          // now we do the actual summation
+          if (!needs_sumtosize_concat_dim && !needs_sumtosize_other_dim) {
+            // nothing to do
+            return raw_outputs[ooffset];
+          } else if (!needs_sumtosize_concat_dim) {
+            // If we don't need to change the concat dim, we can
+            // insert a dimensions so that each tensor has
+            // one index in the new dimension.
+            // Then we can use one sum_to call to do the sum_to_size
+            const auto& ro = raw_outputs[ooffset];
+            size_to_sum_to.erase(
+                size_to_sum_to.begin(), size_to_sum_to.begin() + dims_to_go);
+            return at::sum_to(ro, size_to_sum_to);
```

Comment:
This is confusing. Why are we handling it all manually? I thought that if we have a concat that actually need those `SumToSize`s, then we could generate a fallback graph that excludes it from the fusion group, and run this. It would have a smaller `FusionGroup` inside, which would take care of running the pointwise part (possibly with all the input checks skipped), and the the `STS`es + `aten::cat` that's interpreted as usual! That would be so much simpler than this huge thing!

Reply:
I have reservations about that ending up slower, but I wouldn't know. Clearly the current approach is not too likeable.

Reply:
So I thought about it more and you would need to
1. either detect whether you need the STS before concat or STS to a concatenated area by default (so you'd still have the dance around changing concat dims)
2. `sum_to_out` (which is not unreasonable)
3. rewrite `sts + fusedconcat` to `calculate shape and offsets + empty + sts_out`.
By the time we do 3, my feeling is that we'd have the same I implemented here, just in tensorflow style "define a graph and then run it" instead of imperative.